### ISSUE TITLE
add functionality to the circuit breakers 

### ISSUE THREAD

**Issue by gilbertohasnofb** [Juanvvc/c172p-detailed#80]
As the user tigert posted in the forum thread:

"[...] you can see the circuit breakers in the diagram [see link below], those are listed with their destinations, so they can also be modelled correctly. Breakers are basically current-limited switches that pop open when the rated current is exceeded so you dont melt the wires and get a fire if some equipment breaks. You can also pull them from the knob by hand on some and you can push them back if you suspect the problem is solved."

http://s18.postimg.org/n12yt1qk9/Screenshot_from_2015_03_13_17_33_45.png

**Comment by gilbertohasnofb** [comment]
@Juanvvc wrote in the forum: "Breakers pushed. Not all breakers are functional yet. The master and flaps breakers do nothing yet. The pitot heat breaker does nothing, and probably will do nothing forever"

Also, I believe the interior lights breaker should affect instrument lights, and the turn coordinator breaker isn't working either.

**Comment by Juanvvc** [comment]
Master and flaps breakers were already functional later that night :) Yes, interior lights breaker is not functional yet. The turn coordinator breaker did work when I tested it: remove the breaker with master OFF, switch master ON and the turn coordinator doesn't move. Insert the breaker and the tc starts moving.

The non-funcional breakers are interior light breaker and iirc, some of the exterior lights. I think landing light breaker does work correctly.

**Comment by gilbertohasnofb** [comment]
> pull the breaker with master OFF, switch master ON and the turn coordinator doesn't move. Push bacj the breaker and the tc starts moving.

Thanks, that's certainly working, as well as everything else you mentioned.

**Comment by gilbertohasnofb** [comment]
@Juanvvc I was thinking about this a little bit more and I start to wonder why the turn coordinator does not immediately stop to work when the breaker is pulled. According to the schematics posted in the initial comment, the breaker is the only connection between the TC to the primary bus, and surely pulling it out should make the TC inoperable, shouldn't it?

**Comment by Juanvvc** [comment]
The turn coordinator is one of the main instruments in Flightgear and it is implemented in C++, not in XML or Nasal. The internal gyro decays 0.5% every second, so it takes 200 seconds to stop completely once powered. This is more or less realistic and anyway we cannot do anything to change it except modelling a new turn coordinator in Nasal. This is not going to happen.

**Comment by gilbertohasnofb** [comment]
I see, thanks for the explanation once again, Juan! I didn't know that the TC would take some time to start being unreliable once the breaker was pulled (I thought it would be fast, silly me), and now I understand that your instructions about pulling the breaker before turning the master ON was just to show that it was working properly (since by doing so, the TC will remain in the wrong position it was before starting the plane). I tested now the breaker in flight, and indeed after some short time it already starts being unreliable, so all is perfect!

**Comment by Juanvvc** [comment]
The turn coordinator, the heading indicator and the attitude indicator use internally gyroscopes. These gyroscopes are powered by the vacuum system (the HI and the AI) or the electric system (the TC), in case one of the systems fails you can still rely on the other.

Hence, the HI and the AI are unreliable if the throttle is closed for some minutes (the vacuum system depends on the RPMs) and the TC is unreliable if the electric system is off for some minutes.

**Comment by gilbertohasnofb** [comment]
Ok, so it's all more than clear now! Thanks a lot for the patience to explain these things to me, I really appreciate it, Juan.

**Comment by gilbertohasnofb** [comment]
Are there any non functioning breakers? (I dont have FG on my netbook now to test it). If so, let's close this.

**Comment by onox** [comment]
* Radio stack does depend on master breaker, but not interior lights breaker and radio breaker.
* Unsure: turn coordinator and transponder breaker.
* Display of autopilot does not light up again after re-enabling the autopilot breaker.

**Comment by onox** [comment]
@Juanvvc @tigert If you reduce the instrument lighting, does that also reduce the lighting of the digits of the radio's? And does the lighting of the radio's depend on the interior lights circuit breaker?

**Comment by onox** [comment]
@gilbertohasnofb Are you able to answer these questions?

**Comment by gilbertohasnofb** [comment]
> If you reduce the instrument lighting, does that also reduce the lighting of the digits of the radio's?

I don't think so, because during the day you want the instrument lightning to be at 0%, but you want the radios to display information.

> And does the lighting of the radio's depend on the interior lights circuit breaker?

Therefore I'd say that it doesn't depend on it.

**Comment by onox** [comment]
OK. Should the radio's still depend on the radio breaker? :stuck_out_tongue_winking_eye: I think that wasn't the case when I tested it.

**Comment by gilbertohasnofb** [comment]
> Should the radio's still depend on the radio breaker?

I am 99.9% confident that that is a yes :wink: But the thing is we have 5 radio circuit breakers and one of the AP, and I think that each one controls a single part of the radio. We have to ask someone else which one does what.

**Comment by onox** [comment]
Is the radio breaker sort of a master breaker for the radio's then?

**Comment by gilbertohasnofb** [comment]
Which radio breaker? There are 5 of them numbered 1-5, and I don't see any "master radio breaker".

**Comment by onox** [comment]
Sorry @gilbertohasnofb, I checked `Nasal/electrical.nas` and the breakers are connected to the 5 sets of instruments (audio panel, comm1/nav1, comm2/nav2, transponder, and adf+dme). I was under the impression that radio1..radio4 were connected to comm1, nav1, comm2, and nav2.

**Comment by gilbertohasnofb** [comment]
No problems!

**Comment by onox** [comment]
Do you know if the transponder uses `/systems/electrical/outputs/transponder`?

**Comment by gilbertohasnofb** [comment]
:ok_hand:

**Comment by onox** [comment]
Turn coordinator seems to be ok as well, so we can close this issue.

**Comment by gilbertohasnofb** [comment]
There is still a little problem with Radio 1 braker: it is supposes to shut down the radio, but try moving the rightmost toggle down (to test position) and you'll see it illuminates the runway markers lights. I didn't try landing with this breaker off, but my guess is that the markers will also be working in that scenario.

**Comment by onox** [comment]
radio1 breaker is connected to the audio panel;

```javascript
496     # Audio Panel 1 Power
497     if ( getprop("/controls/circuit-breakers/radio1") ) {
498       setprop("/systems/electrical/outputs/audio-panel[0]", bus_volts);
499     } else {
500       setprop("/systems/electrical/outputs/audio-panel[0]", 0.0);
501     }
```

**Comment by onox** [comment]
I guess you're talking about radio2?

**Comment by onox** [comment]
@gilbertohasnofb The audio panel is the marker-beacon instrument and it uses `/systems/electrical/outputs/nav[0]` for bus power.

**Comment by gilbertohasnofb** [comment]
> You probably meant the audio panel and marker beacons.

Yes! Sorry, I expressed myself quite badly.

I still think there is something wrong with the marker beacons. Even with the battery off and the the avionics switch off, they are still being illuminated when I click on the test toggle!

**Comment by onox** [comment]
Maybe the three animations in `Models/Interior/Panel/Instruments/kma20/kma20.xml` on lines 96 to 130 need to be updated then.

**Comment by onox** [comment]
@Juanvvc @tigert Can you guys find out how much volt the audio panel needs? The current C++ says at least 1.0 V. That is way too low :stuck_out_tongue_winking_eye:

**Comment by onox** [comment]
@gilbertohasnofb Please test PR #301. Note that the marker-beacon instrument (used by the audio panel) listens to `/systems/electrical/outputs/nav[0]`. See https://github.com/Juanvvc/c172p-detailed/issues/80#issuecomment-111865696

**Comment by onox** [comment]
I need to add some conditions to `c172-sound.xml`.

**Comment by onox** [comment]
@gilbertohasnofb Does the stall sound depend on the circuit breaker for the audio panel? There's also a KAP140Beep sound. I think that one should depend on the circuit breaker for the audio panel as well.

**Comment by onox** [comment]
Note that `Nasal/kma20.nas` controls various volumes, but the C++ instruments do not care about the voltage of `/systems/electrical/outputs/audio-panel[0]`.

Should various sounds (marker beacon, afd, dme, comms) depend on whether the audio panel has power or not?

**Comment by gilbertohasnofb** [comment]
@onox 

> Please test PR #301

Will do that in a short while!

> Does the stall sound depend on the circuit breaker for the audio panel?

I am not sure, but I believe the stall horn is a simple mechanical device on the wing and the sound comes directly from it. So no electronics, wires, amplifiers, speakers, etc. 

> Should various sounds (marker beacon, afd, dme, comms) 

Most probably yes. But maybe @Juanvvc and @tigert could clarify this better.

**Comment by onox** [comment]
> Most probably yes. But maybe @Juanvvc and @tigert could clarify this better.

In that case I think we should create a separate issue about need of the kma20 to control the actual volume of various instruments depending on its voltage.

**Comment by Juanvvc** [comment]
The stall horn is mechanical, and it works without any power. One of the things you have to do during the external inspection is aspiring from the stall detector (a hole in the wing) and check if the stall horn makes its sound. This is done without power. Not all aircraft use mechanical detectors: Cessna and Diamond aircraft usually have mechanical stall alarms. Piper usually has electrical stall alarms.

All sounds in the comm panel depend on the audio panel: vor, adf, comms. The audio panel is the amplifier of all audio channels, and also the device that selects if COMM1, COMM2, VOR1, VOR2 and ADF (or all of them together!) are connected to the ear speakers/mic or the cabin speakers/mic.

**Comment by onox** [comment]
I have opened a separate issue for the KMA20 instrument. Closing this issue.
